File: avalon-modules/avalon-authorization/src/main/java/io/github/lishangbu/avalon/authorization/service/impl/MenuServiceImpl.java
package io.github.lishangbu.avalon.authorization.service.impl;

import io.github.lishangbu.avalon.authorization.entity.Menu;
import io.github.lishangbu.avalon.authorization.model.MenuTreeNode;
import io.github.lishangbu.avalon.authorization.repository.MenuRepository;
import io.github.lishangbu.avalon.authorization.service.MenuService;
import io.github.lishangbu.avalon.web.util.TreeUtils;
import java.util.ArrayDeque;
import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.function.BiConsumer;
import java.util.function.Function;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Example;
import org.springframework.data.domain.ExampleMatcher;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.CollectionUtils;
import org.springframework.util.StringUtils;

/// 菜单服务接口实现
///
/// 提供根据角色代码构建菜单树的实现
///
/// @author lishangbu
/// @since 2025/9/19
@Slf4j
@Service
@RequiredArgsConstructor
public class MenuServiceImpl implements MenuService {
    private static final ExampleMatcher MENU_QUERY_MATCHER =
            ExampleMatcher.matching()
                    .withIgnoreNullValues()
                    .withMatcher("icon", ExampleMatcher.GenericPropertyMatchers.contains())
                    .withMatcher("key", ExampleMatcher.GenericPropertyMatchers.contains())
                    .withMatcher("label", ExampleMatcher.GenericPropertyMatchers.contains())
                    .withMatcher("path", ExampleMatcher.GenericPropertyMatchers.contains())
                    .withMatcher("name", ExampleMatcher.GenericPropertyMatchers.contains())
                    .withMatcher("redirect", ExampleMatcher.GenericPropertyMatchers.contains())
                    .withMatcher("component", ExampleMatcher.GenericPropertyMatchers.contains());

    private static final Sort MENU_TREE_SORT =
            Sort.by(Sort.Order.asc("sortingOrder"), Sort.Order.asc("id"));

    private final MenuRepository menuRepository;

    @Override
    public List<MenuTreeNode> listMenuTreeByRoleCodes(List<String> roleCodes) {
        if (CollectionUtils.isEmpty(roleCodes)) {
            return Collections.emptyList();
        }
        List<Menu> menus = menuRepository.findAllByRoleCodes(roleCodes);
        log.debug("根据角色编码获取到 [{}] 条菜单记录", menus == null ? 0 : menus.size());
        return buildTreeFromMenus(menus);
    }

    @Override
    public List<MenuTreeNode> listAllMenuTree(Menu menu) {
        Menu condition = normalizeCondition(menu);
        List<Menu> allMenus = menuRepository.findAllByOrderBySortingOrderAscIdAsc();
        if (CollectionUtils.isEmpty(allMenus)) {
            return Collections.emptyList();
        }
        if (!hasQueryCondition(condition)) {
            return buildTreeFromMenus(allMenus);
        }
        List<Menu> matchedMenus =
                menuRepository.findAll(Example.of(condition, MENU_QUERY_MATCHER), MENU_TREE_SORT);
        if (CollectionUtils.isEmpty(matchedMenus)) {
            return Collections.emptyList();
        }
        Map<Long, Menu> menuById =
                allMenus.stream()
                        .filter(menuItem -> menuItem.getId() != null)
                        .collect(Collectors.toMap(Menu::getId, Function.identity()));
        Map<Long, List<Menu>> childrenByParentId =
                allMenus.stream()
                        .filter(menuItem -> menuItem.getId() != null)
                        .collect(Collectors.groupingBy(Menu::getParentId));

        Set<Long> includedIds = new LinkedHashSet<>();
        for (Menu matchedMenu : matchedMenus) {
            if (matchedMenu.getId() == null) {
                continue;
            }
            collectAncestors(menuById, matchedMenu.getId(), includedIds);
            collectDescendants(matchedMenu.getId(), childrenByParentId, includedIds);
        }
        if (includedIds.isEmpty()) {
            return Collections.emptyList();
        }
        List<Menu> filteredMenus =
                allMenus.stream()
                        .filter(
                                menuItem ->
                                        menuItem.getId() != null
                                                && includedIds.contains(menuItem.getId()))
                        .toList();
        return buildTreeFromMenus(filteredMenus);
    }

    @Override
    public Optional<Menu> getById(Long id) {
        return menuRepository.findById(id);
    }

    @Override
    @Transactional(rollbackFor = Exception.class)
    public Menu save(Menu menu) {
        return menuRepository.save(menu);
    }

    @Override
    @Transactional(rollbackFor = Exception.class)
    public Menu update(Menu menu) {
        return menuRepository.save(menu);
    }

    @Override
    @Transactional(rollbackFor = Exception.class)
    public void removeById(Long id) {
        menuRepository.deleteById(id);
    }

    private static void collectAncestors(
            Map<Long, Menu> menuById, Long startId, Set<Long> includedIds) {
        Long currentId = startId;
        while (currentId != null && includedIds.add(currentId)) {
            Menu current = menuById.get(currentId);
            if (current == null) {
                break;
            }
            currentId = current.getParentId();
        }
    }

    private static void collectDescendants(
            Long startId, Map<Long, List<Menu>> childrenByParentId, Set<Long> includedIds) {
        if (startId == null) {
            return;
        }
        ArrayDeque<Long> deque = new ArrayDeque<>();
        Set<Long> visited = new LinkedHashSet<>();
        deque.push(startId);
        while (!deque.isEmpty()) {
            Long currentId = deque.pop();
            if (!visited.add(currentId)) {
                continue;
            }
            includedIds.add(currentId);
            for (Menu child : childrenByParentId.getOrDefault(currentId, Collections.emptyList())) {
                if (child.getId() != null) {
                    deque.push(child.getId());
                }
            }
        }
    }

    private static Menu normalizeCondition(Menu menu) {
        if (menu == null) {
            return null;
        }
        Menu normalized = new Menu();
        normalized.setId(menu.getId());
        normalized.setParentId(menu.getParentId());
        normalized.setDisabled(menu.getDisabled());
        normalized.setExtra(menu.getExtra());
        normalized.setIcon(trimToNull(menu.getIcon()));
        normalized.setKey(trimToNull(menu.getKey()));
        normalized.setLabel(trimToNull(menu.getLabel()));
        normalized.setShow(menu.getShow());
        normalized.setPath(trimToNull(menu.getPath()));
        normalized.setName(trimToNull(menu.getName()));
        normalized.setRedirect(trimToNull(menu.getRedirect()));
        normalized.setComponent(trimToNull(menu.getComponent()));
        normalized.setSortingOrder(menu.getSortingOrder());
        normalized.setPinned(menu.getPinned());
        normalized.setShowTab(menu.getShowTab());
        normalized.setEnableMultiTab(menu.getEnableMultiTab());
        return normalized;
    }

    private static String trimToNull(String value) {
        return StringUtils.hasText(value) ? value.trim() : null;
    }

    private static boolean hasQueryCondition(Menu menu) {
        return menu != null
                && (menu.getId() != null
                        || menu.getParentId() != null
                        || menu.getDisabled() != null
                        || menu.getExtra() != null
                        || menu.getIcon() != null
                        || menu.getKey() != null
                        || menu.getLabel() != null
                        || menu.getShow() != null
                        || menu.getPath() != null
                        || menu.getName() != null
                        || menu.getRedirect() != null
                        || menu.getComponent() != null
                        || menu.getSortingOrder() != null
                        || menu.getPinned() != null
                        || menu.getShowTab() != null
                        || menu.getEnableMultiTab() != null);
    }

    /// 将权限实体列表转换为树节点并构建树结构
    ///
    /// - 处理空集合，返回不可变空列表
    /// - 将 Menu 映射为 MenuTreeNode
    /// - 使用通用的 [TreeUtils] 构建树
    ///
    /// @param menus 权限实体列表，允许为 null
    /// @return 树结构的 MenuTreeNode 列表，永远不返回 null
    /// @see TreeUtils#buildTree(List, Function, Function, BiConsumer)
    private List<MenuTreeNode> buildTreeFromMenus(List<Menu> menus) {
        if (CollectionUtils.isEmpty(menus)) {
            return Collections.emptyList();
        }

        List<MenuTreeNode> treeNodes = menus.stream().map(MenuTreeNode::new).toList();

        return TreeUtils.buildTree(
                treeNodes,
                MenuTreeNode::getId,
                MenuTreeNode::getParentId,
                MenuTreeNode::setChildren);
    }
}

File: avalon-modules/avalon-authorization/src/main/kotlin/io/github/lishangbu/avalon/authorization/service/impl/MenuServiceImpl.kt
package io.github.lishangbu.avalon.authorization.service.impl

import io.github.lishangbu.avalon.authorization.entity.Menu
import io.github.lishangbu.avalon.authorization.entity.Role
import io.github.lishangbu.avalon.authorization.entity.dto.MenuSpecification
import io.github.lishangbu.avalon.authorization.entity.dto.MenuTreeView
import io.github.lishangbu.avalon.authorization.entity.dto.MenuView
import io.github.lishangbu.avalon.authorization.entity.dto.SaveMenuInput
import io.github.lishangbu.avalon.authorization.entity.dto.UpdateMenuInput
import io.github.lishangbu.avalon.authorization.repository.AuthorizationFetchers
import io.github.lishangbu.avalon.authorization.repository.MenuRepository
import io.github.lishangbu.avalon.authorization.repository.RoleRepository
import io.github.lishangbu.avalon.authorization.service.MenuService
import io.github.lishangbu.avalon.jimmer.support.readOrNull
import org.babyfish.jimmer.sql.ast.mutation.SaveMode
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/**
 * 菜单服务实现
 *
 * 负责菜单查询与菜单树构建
 *
 * @author lishangbu
 * @since 2025/9/19
 */
@Service
class MenuServiceImpl(
    /** 菜单仓储 */
    private val menuRepository: MenuRepository,
    /** 角色仓储 */
    private val roleRepository: RoleRepository,
) : MenuService {
    /** 根据角色编码列表查询菜单树列表 */
    override fun listMenuTreeByRoleCodes(roleCodes: List<String>): List<MenuTreeView> {
        if (roleCodes.isEmpty()) {
            return emptyList()
        }
        val menus = menuRepository.listViewsByRoleCodes(roleCodes)
        log.debug("根据角色编码获取到 [{}] 条菜单记录", menus.size)
        return buildRoleMenuTree(menus)
    }

    /** 查询全部菜单树列表 */
    override fun listTree(): List<MenuTreeView> = menuRepository.listTreeViews()

    /** 按条件查询菜单列表 */
    override fun listByCondition(specification: MenuSpecification): List<MenuView> = menuRepository.listViews(specification)

    /** 按 ID 查询菜单 */
    override fun getById(id: Long): MenuView? = menuRepository.loadViewById(id)

    /** 保存菜单 */
    @Transactional(rollbackFor = [Exception::class])
    override fun save(command: SaveMenuInput): MenuView {
        val menu = command.toEntity()
        validateParent(menu)
        return menuRepository.save(menu, SaveMode.INSERT_ONLY).let(::reloadView)
    }

    /** 更新菜单 */
    @Transactional(rollbackFor = [Exception::class])
    override fun update(command: UpdateMenuInput): MenuView {
        val menu = command.toEntity()
        validateParent(menu)
        return menuRepository.save(menu).let(::reloadView)
    }

    /** 按 ID 删除菜单 */
    @Transactional(rollbackFor = [Exception::class])
    override fun removeById(id: Long) {
        if (menuRepository.hasChildren(id)) {
            throw IllegalStateException("请先删除子菜单后再删除当前菜单")
        }
        detachMenuFromRoles(id)
        menuRepository.deleteById(id)
    }

    private fun validateParent(menu: Menu) {
        val menuId = menu.readOrNull { id }
        val parentId = menu.readOrNull { parent }?.id ?: return

        if (menuId != null && parentId == menuId) {
            throw IllegalArgumentException("父菜单不能选择自身")
        }

        val parent =
            menuRepository.findNullable(parentId, AuthorizationFetchers.MENU)
                ?: throw IllegalArgumentException("父菜单不存在")

        if (menuId != null) {
            validateNoCircularReference(menuId, parent)
        }
    }

    private fun validateNoCircularReference(
        menuId: Long,
        parent: Menu,
    ) {
        var currentId: Long? = parent.id
        val visited = mutableSetOf<Long>()

        while (currentId != null) {
            if (currentId == menuId) {
                throw IllegalStateException("父菜单不能选择当前菜单或其子菜单")
            }
            if (!visited.add(currentId)) {
                throw IllegalStateException("菜单层级存在循环引用")
            }
            currentId =
                menuRepository
                    .findNullable(currentId, AuthorizationFetchers.MENU)
                    ?.readOrNull { parent }
                    ?.id
        }
    }

    private fun detachMenuFromRoles(menuId: Long) {
        val roles =
            roleRepository
                .listWithMenus(null)
                .filter { role -> role.menus.any { menu -> menu.id == menuId } }

        roles.forEach { role ->
            val remainedMenus = role.menus.filterNot { menu -> menu.id == menuId }
            roleRepository.save(
                Role(role) {
                    menus = remainedMenus
                },
            )
        }
    }

    private fun buildRoleMenuTree(menus: List<MenuView>): List<MenuTreeView> {
        if (menus.isEmpty()) {
            return emptyList()
        }
        val ids = menus.mapTo(linkedSetOf()) { it.id }
        val childrenByParentId = menus.groupBy { it.parentId }

        fun toTree(
            node: MenuView,
            path: Set<String> = emptySet(),
        ): MenuTreeView {
            if (node.id in path) {
                throw IllegalStateException("菜单层级存在循环引用")
            }
            val children =
                childrenByParentId[node.id]
                    .orEmpty()
                    .map { child -> toTree(child, path + node.id) }
            return MenuTreeView(
                id = node.id,
                parentId = node.parentId,
                disabled = node.disabled,
                extra = node.extra,
                icon = node.icon,
                key = node.key,
                label = node.label,
                show = node.show,
                path = node.path,
                name = node.name,
                redirect = node.redirect,
                component = node.component,
                sortingOrder = node.sortingOrder,
                pinned = node.pinned,
                showTab = node.showTab,
                enableMultiTab = node.enableMultiTab,
                children = children,
            )
        }

        return menus
            .filter { menu -> menu.parentId == null || menu.parentId !in ids }
            .map(::toTree)
    }

    companion object {
        /** 日志记录器 */
        private val log: Logger = LoggerFactory.getLogger(MenuServiceImpl::class.java)
    }

    private fun reloadView(menu: Menu): MenuView = requireNotNull(menuRepository.loadViewById(menu.id)) { "未找到 ID=${menu.id} 对应的菜单" }
}
